src/main/java/org/folio/marccat/business/cataloguing/authority/AuthorityCatalog.java

```diff
@@ -126,26 +153,87 @@ protected CorrelationKey getCorrelationKeyByControlfield() {
   @Override
   public Tag getNewTag(CatalogItem item, int category, CorrelationValues correlationValues) {
     Tag tag = (Tag) getTagFactory().create(category);
-    tag = (Tag) setItemIfNecessary(item, tag);
 
-    if (correlationValues != null) {
-      if (tag.correlationChangeAffectsKey(correlationValues)) {
-        if (tag instanceof HeaderField) {
-          tag = (Tag) getFixedFieldFactory().create(correlationValues.getValue(1));
-          tag = (Tag) setItemIfNecessary(item, tag);
-          tag.setCorrelation(1, correlationValues.getValue(1));
+    if (correlationValues != null && tag.correlationChangeAffectsKey(correlationValues)) {
+      if (tag instanceof HeaderField) {
+        tag = (Tag) getFixedFieldFactory().create(correlationValues.getValue(1));
+      } else if (tag instanceof AuthorityHeadingTag || tag instanceof AuthorityReferenceTag) {
+        Integer refTypePosition = correlationValues.getLastUsedPosition();
+        int refType;
+
+        if (refTypePosition != null) {
+          refType = correlationValues.getValue(refTypePosition.intValue());
         } else {
-          tag.setCorrelationValues(correlationValues);
+          throw new ModMarccatException("attempt to create reference tag with no reference type specified");
         }
```

Comment:
There are a lot of branches here... in some cases you can reduce some of that via gates, e.g. 
```suggestion
        if (refTypePosition == null) {
          throw new ModMarccatException("attempt to create reference tag with no reference type specified");
        }
        int refType = correlationValues.getValue(refTypePosition.intValue());
```

Reply:
Thanks a lot.